samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/util/SamigoExpressionParser.java

```diff
@@ -35,6 +35,10 @@
   public static String[] oldSamigoParserVars = {"SIN", "COS", "TAN", "ASIN", "ACOS", "ATAN", "ABS", "EXP", "SGN", "SQRT", "LOG10", "LN", "PI", "E", "SIGN", "LOG", "log"};
   public static String[] newSamigoParserVars = {"sin", "cos", "tan", "asin", "acos", "atan", "abs", "exp", "sgn", "sqrt", "log10", "ln", "pi", "e", "sgn",  "ln",  "ln"};
 
+  // Special case for factorial replacement
+  public static String OLD_FACTORIAL_PATTERN = "(?i)factorial\\((\\d+)\\)";
```

Comment:
I believe this matches digits only? I see some occurrences in our database like:
FACTORIAL({N} - {n})
But the nested expression can be anything - it could involve balancing parentheses, etc., which is too complex for Regex. So I think this is good - keep the regex simple to solve 80% of the cases; 20% can be fixed in the DB as needed.